Enabled torch.mm and torch.mv for bfloat16 on CPU

izdeby opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#24224 Enabled torch.mm and torch.mv for bfloat16 on CPU**

Differential Revision: [D16779996](https://our.internmc.facebook.com/intern/diff/D16779996)

----------

Enabled torch.mm and torch.mv for bfloat16 dtype. In order to do that, these methods have to be enabled for bfloat16.
- mul
- div
- addmm
- addmv
- addr

Also, overloaded arithmetic ops for BFloat16 including some bitwise ops such as '|', '&' and '^'

Tested via unit tests

izdeby committed:
Enabled torch.mm and torch.mv for bfloat16

izdeby committed:
Update on "[WIP] Enabled torch.mm and torch.mv for bfloat16"


Differential Revision: [D16779996](https://our.internmc.facebook.com/intern/diff/D16779996)

VitalyFedyunin commented on test/test_torch.py:
```diff
@@ -377,7 +383,7 @@ def test_addmm(self):
                     for k in range(50):
                         res2[i, j] += m1[i, k] * m2[k, j]
-            self.assertEqual(res1, res2)
+            self.assertEqual(res1, res2, _prec)
```
Usually underscore notation means unused variable, so you might want to rename it to `prec`.

VitalyFedyunin reviewed (commented):
Will merge conflict with #23492 ( cc @pbelevich )

ezyang reviewed (changes requested):
yeah, don't use `_prec` please

izdeby committed:
Update on "Enabled torch.mm and torch.mv for bfloat16 on CPU"


Differential Revision: [D16779996](https://our.internmc.facebook.com/intern/diff/D16779996)

----------

Enabled torch.mm and torch.mv for bfloat16 dtype. In order to do that, these methods have to be enabled for bfloat16.
- mul
- div
- addmm
- addmv
- addr

Also, overloaded arithmetic ops for BFloat16 including some bitwise ops such as '|', '&' and '^'

Tested via unit tests

gchanan commented on test/test_torch.py:
```diff
@@ -324,8 +326,9 @@ def test_addmv(self):
         types = {
             'torch.DoubleTensor': 1e-8,
             'torch.FloatTensor': 1e-4,
+            'torch.BFloat16Tensor': 0.4,
```
use the same format please.

gchanan commented on test/test_torch.py:
```diff
@@ -348,14 +352,16 @@ def test_addmv(self):
         types = {
             'torch.DoubleTensor': 1e-8,
             'torch.FloatTensor': 1e-4,
+            'torch.BFloat16Tensor': 0.3,
```
use the  same format please.

izdeby committed:
Update on "Enabled torch.mm and torch.mv for bfloat16 on CPU"


Differential Revision: [D16779996](https://our.internmc.facebook.com/intern/diff/D16779996)

----------

Enabled torch.mm and torch.mv for bfloat16 dtype. In order to do that, these methods have to be enabled for bfloat16.
- mul
- div
- addmm
- addmv
- addr

Also, overloaded arithmetic ops for BFloat16 including some bitwise ops such as '|', '&' and '^'

Tested via unit tests

izdeby committed:
Update on "Enabled torch.mm and torch.mv for bfloat16 on CPU"


Differential Revision: [D16779996](https://our.internmc.facebook.com/intern/diff/D16779996)

----------

Enabled torch.mm and torch.mv for bfloat16 dtype. In order to do that, these methods have to be enabled for bfloat16.
- mul
- div
- addmm
- addmv
- addr

Also, overloaded arithmetic ops for BFloat16 including some bitwise ops such as '|', '&' and '^'

Tested via unit tests
